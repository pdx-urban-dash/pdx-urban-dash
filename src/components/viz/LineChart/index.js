import React from 'react';
import PropTypes from 'prop-types';
import { createClassFromSpec } from 'react-vega';

import { injectPropsIntoSchema } from '../../../utils/vegaUtils';
import { getDefaultColor, getTrendLineFunc } from '../../../utils/vizUtils';
import { colors, sizes } from '../../../constants';
import lineChartSchema from './schema';
import './styles.scss';

import Legend from '../shared/Legend';
import ChartWrapper from '../shared/ChartWrapper';


export const buildData = (data) => {
  const retData = [];
  let currentTrendLineNumber = data.length;
  data.forEach((dataSet, idx) => {
    dataSet.values.forEach(point => retData.push({ x: point.x, y: point.y, c: idx }));
    if (dataSet.showTrendLine) {
      const trendFunc = getTrendLineFunc(dataSet.values);
      dataSet.values.forEach(
        ({ x }) => retData.push({ x, y: trendFunc(x), c: currentTrendLineNumber }),
      );
      currentTrendLineNumber += 1;
    }
  });
  return retData;
};

export const getColors = (data) => {
  let numTrendLines = 0;
  const mColors = data.map((elem, idx) => {
    if (elem.showTrendLine) numTrendLines += 1;
    return elem.lineColor || getDefaultColor(idx);
  });
  for (let i = 0; i < numTrendLines; i += 1) {
    mColors.push(colors.trendLine);
  }
  return mColors;
};

const LineChart = (
  {
    data,
    title,
    className,
    colorScheme,
  },
) => {
  const colorRange = getColors(data);
  const Graph = createClassFromSpec(
    injectPropsIntoSchema(
      {
        $dataValues: {
          name: 'values',
          value: buildData(data),
        },
        $labelColor: {
          name: 'labelColor',
          value: colorScheme === 'light' ? colors.fontDark : colors.fontLight,
        },
        $labelFontSize: {
          name: 'labelFontSize',
          value: parseInt(sizes.fontSM.slice(0, -2), 10),
        },
        $colorRange: {
          name: 'range',
          value: colorRange,
        },
      },
      lineChartSchema,
    ),
  );
  return (
    <ChartWrapper title={title} className={className}>
      <Graph />
      <Legend
        series={data.map((elem, idx) => ({
          title: elem.dataSetName,
          description: elem.description,
          color: elem.lineColor || getDefaultColor(idx),
        }))}
      />
    </ChartWrapper>
  );
};

LineChart.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({
    dataSetName: PropTypes.string,
    lineColor: PropTypes.string,
    description: PropTypes.string,
    showTrendLine: PropTypes.bool,
    values: PropTypes.arrayOf(PropTypes.shape({
      x: PropTypes.number.isRequired,
      y: PropTypes.number.isRequired,
    })),
  })).isRequired,
  title: PropTypes.string,
  className: PropTypes.string,
  colorScheme: PropTypes.oneOf(['light', 'dark']),
};

LineChart.defaultProps = {
  title: 'chart title',
  className: '',
  colorScheme: 'light',
};

export default LineChart;
