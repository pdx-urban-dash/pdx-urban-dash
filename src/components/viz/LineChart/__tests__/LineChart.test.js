import React from 'react';
import { shallow } from 'enzyme';
import LineChart from '../index';


describe('<LineChart />', () => {
  test('snapshot', () => {
    const wrapper = shallow(
      <LineChart
        data={[
          {
            dataSetName: 'first',
            description: 'this is first',
            lineColor: 'red',
            showTrendLine: true,
            values: [
              { x: 0, y: 10 },
              { x: 1, y: 20 },
              { x: 2, y: 15 },
              { x: 3, y: 40 },
            ],
          },
          {
            dataSetName: 'second',
            lineColor: 'blue',
            description: 'this is second',
            values: [
              { x: 0, y: 20 },
              { x: 1, y: 30 },
              { x: 2, y: 19 },
              { x: 3, y: 50 },
            ],
          },
        ]}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
