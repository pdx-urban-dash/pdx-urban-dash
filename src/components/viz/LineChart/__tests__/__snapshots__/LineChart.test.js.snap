// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LineChart /> snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LineChart
    className=""
    colorScheme="light"
    data={
      Array [
        Object {
          "dataSetName": "first",
          "description": "this is first",
          "lineColor": "red",
          "showTrendLine": true,
          "values": Array [
            Object {
              "x": 0,
              "y": 10,
            },
            Object {
              "x": 1,
              "y": 20,
            },
            Object {
              "x": 2,
              "y": 15,
            },
            Object {
              "x": 3,
              "y": 40,
            },
          ],
        },
        Object {
          "dataSetName": "second",
          "description": "this is second",
          "lineColor": "blue",
          "values": Array [
            Object {
              "x": 0,
              "y": 20,
            },
            Object {
              "x": 1,
              "y": 30,
            },
            Object {
              "x": 2,
              "y": 19,
            },
            Object {
              "x": 3,
              "y": 50,
            },
          ],
        },
      ]
    }
    title="chart title"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <a />,
        <Legend
          series={
            Array [
              Object {
                "color": "red",
                "description": "this is first",
                "title": "first",
              },
              Object {
                "color": "blue",
                "description": "this is second",
                "title": "second",
              },
            ]
          }
        />,
      ],
      "className": "",
      "title": "chart title",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "series": Array [
            Object {
              "color": "red",
              "description": "this is first",
              "title": "first",
            },
            Object {
              "color": "blue",
              "description": "this is second",
              "title": "second",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <a />,
          <Legend
            series={
              Array [
                Object {
                  "color": "red",
                  "description": "this is first",
                  "title": "first",
                },
                Object {
                  "color": "blue",
                  "description": "this is second",
                  "title": "second",
                },
              ]
            }
          />,
        ],
        "className": "",
        "title": "chart title",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "series": Array [
              Object {
                "color": "red",
                "description": "this is first",
                "title": "first",
              },
              Object {
                "color": "blue",
                "description": "this is second",
                "title": "second",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
